Add support for dynamic workspace files based on mustache templates (#7091)
PrairieLearn/PrairieLearn: docs/workspaces/index.md

</pl-submission-panel>
```

## Creating files in the workspace home directory

Workspace questions can optionally include a `workspace/` subdirectory within the regular [PrairieLearn question directory structure](../question.md#directory-structure). If this `workspace/` subdirectory exists, its contents will be copied into the home directory of the student's workspace container, as configured in the `home` setting in `info.json`.

> Questions using workspaces can also be randomized, i.e., include files that contain random and dynamic content. This may be done in two ways: using mustache-based template files or using [the `server.py` file in the question directory](../question.md#question-serverpy). For template files, a workspace question can optionally include a `workspaceTemplates/` subdirectory within the regular question directory structure. The contents will be copied into the home directory of the student's workspace container, as with the `workspace/` directory. However, all files within this directory may include mustache tags (e.g., `{{params.value}}`), which will be replaced with the equivalent values set by `server.py`. For example, if `server.py` sets `data["params"]["starting_value"]` to `17`, then if the file `main.py` inside `workspaceTemplates` has the following content:

Reviewer: Recommendation: support an optional `.mustache` extension on files, and strip off that suffix if it exists. This will help folks like me who get endlessly annoyed when my editor complains about invalid syntax in templated files. In other words, the below file could be `main.py.mustache`, and it would appear to the workspace as `main.py`.

If someone really wants a templated mustache file that is itself a mustache file (that hurts my brain), they could us `main.mustache.mustache`.

Author: Done in 9b3a54b9b

> ```py

Reviewer: ```suggestion
```txt
```

> Questions using workspaces can also be randomized, i.e., include files that contain random and dynamic content. This may be done in two ways: using mustache-based template files or using [the `server.py` file in the question directory](../question.md#question-serverpy). For template files, a workspace question can optionally include a `workspaceTemplates/` subdirectory within the regular question directory structure. The contents will be copied into the home directory of the student's workspace container, as with the `workspace/` directory. However, files within this directory may include mustache tags (e.g., `{{params.value}}`), which will be replaced with the equivalent values set by `server.py`. File names may optionally include the `.mustache` extension, and the file will be renamed before being presented to the student. For example, if `server.py` sets `data["params"]["starting_value"]` to `17`, then if the file `main.py,mustache` inside `workspaceTemplates` has the following content:

Reviewer: ```suggestion
Questions using workspaces can also be randomized, i.e., include files that contain random and dynamic content. This may be done in two ways: using mustache-based template files or using [the `server.py` file in the question directory](../question.md#question-serverpy). For template files, a workspace question can optionally include a `workspaceTemplates/` subdirectory within the regular question directory structure. The contents will be copied into the home directory of the student's workspace container, as with the `workspace/` directory. However, files within this directory may include mustache tags (e.g., `{{params.value}}`), which will be replaced with the equivalent values set by `server.py`. File names may optionally include the `.mustache` extension, and the file will be renamed before being presented to the student. For example, if `server.py` sets `data["params"]["starting_value"]` to `17`, then if the file `main.py.mustache` inside `workspaceTemplates` has the following content:
```

    ]
```

By default, `contents` is expected to be a string in UTF-8 format. To provide binary content, the value must be encoded using base64 or hex, as shown in the example above. In this case, the `encoding` property must also be provided. Only one of `questionFile` and `contents` may be provided. If neither `questionFile` nor `contents` are provided, an empty file is created.

> If a file name is found in both the dynamic list from `_workspace_files` and in the `workspace/` or `workspaceTemplates/` directory, the dynamic content is used. For file names found in both `workspace/` and `workspaceTemplates/`, the template version is used.

Reviewer: Minor suggestion; I find an ordered, bulleted list easier to parse.

```suggestion
If a file appears in multiple locations, the following precedence takes effect (higher precedence first):

- dynamic content from `_workspace_files`
- `workspaceTemplate/` directory
- `workspace/` directory
```
